src/app-layer-imap.c

```diff
@@ -0,0 +1,606 @@
+         *identifier (typically a short alphanumeric string, e.g., A0001, A0002, etc.) called a
+         *"tag".
+         * */
+        uint32_t command_start = 0;
+        if (!(command_start = IMAPTaggedCmd(input, input_len))) {
+            goto not_imap;
+        }
+
+        if (IMAPCmd(input, input_len, command_start)) {
+            goto imap_detected;
+        }
+    }
+
+not_imap:
+    SCLogDebug("Protocol not detected as IMAP.");
+    return ALPROTO_UNKNOWN;
+
+imap_detected:
+    SCLogDebug("Protocol detected as IMAP.");
+    return ALPROTO_IMAP;
+}
+
+/**
+ * \brief Probe the input to see if it looks like imap, checks server responses.
+ *
+ * \retval ALPROTO_IMAP if it looks like imap, otherwise
+ *     ALPROTO_UNKNOWN.
+ */
+static AppProto IMAPProbingParserTc(
+        Flow *f, uint8_t direction, const uint8_t *input, uint32_t input_len, uint8_t *rdir)
+{
+
+    if (input_len < IMAP_MIN_FRAME_LEN) {
+        goto not_imap;
+    }
+
+    uint16_t line_end;
+    int ret = ByteExtractUint16(
+            &line_end, BYTE_BIG_ENDIAN, sizeof(uint16_t), (const uint8_t *)(input + input_len - 2));
+    if (ret == 0) {
+        goto not_imap;
+    }
+
+    if (line_end != 0x0d0a /*CRLF*/) {
+        goto not_imap;
+    }
+
+    /**
+     * after first token there is a single space then
+     * generic status (OK, NO, BAD, PREAUTH, BYE)
+     * server status (ENABLED, CAPABILITY)
+     * mailbox status (LIST, NAMESPACE, STATUS, SEARCH, FLAGS)
+     * mailbox size (EXISTS)
+     * message status (EXPUNGE, FETCH)
+     * */
+    if (IMAPServerResponseGenericStatus(input, input_len)) {
+        goto imap_detected;
+    }
+
+    if (IMAPServerResponseStatus(input, input_len)) {
+        goto imap_detected;
+    }
+
+    if (IMAPServerResponseMailboxStatus(input, input_len)) {
+        goto imap_detected;
+    }
+
+    if (IMAPServerResponseMailboxSize(input, input_len)) {
+        goto imap_detected;
+    }
+
+    if (IMAPServerResponseMessageStatus(input, input_len)) {
+        goto imap_detected;
+    }
+
+not_imap:
+    SCLogDebug("Protocol not detected as IMAP.");
+    return ALPROTO_UNKNOWN;
+
+imap_detected:
+    SCLogDebug("Protocol detected as IMAP.");
+    return ALPROTO_IMAP;
+}
+
+void RegisterIMAPParsers(void)
+{
+    const char *proto_name = "imap";
+
+    if (AppLayerProtoDetectConfProtoDetectionEnabled("tcp", proto_name)) {
+
+        SCLogDebug("IMAP protocol detection is enabled.");
+
+        AppLayerProtoDetectRegisterProtocol(ALPROTO_IMAP, proto_name);
+
+        if (!AppLayerProtoDetectPPParseConfPorts("tcp", IPPROTO_TCP, proto_name, ALPROTO_IMAP,
+                    IMAP_MIN_FRAME_LEN, 0, IMAPProbingParserTs, IMAPProbingParserTc)) {
+            SCLogDebug("No imap app-layer configuration, enabling imap"
+                       " detection on port %s.",
+                    IMAP_DEFAULT_PORT);
+
+            AppLayerProtoDetectPPRegister(IPPROTO_TCP, IMAP_DEFAULT_PORT, ALPROTO_IMAP,
```

Comment:
this only registers expensive probing parsers on the default port 143. Can we use PM+PP logic here? e.g. `AppLayerProtoDetectPMRegisterPatternCSwPP`. This would allow a weak pattern that is port-independent, with the PP then called to validate.